types.ts

```diff
@@ -84,24 +154,94 @@ export interface Field {
     pattern?: string;
     [key: string]: unknown;
   };
+  /** @deprecated. Use `upload` instead */
+  uploads?: string;
+  upload?: string | false;
+  publicPath?: string;
+  options?: Option[];
+  fields?: Field[];
   init?: (field: ResolvedField, content: CMSContent) => void | Promise<void>;
   transform?(value: any, field: ResolvedField): any;
-  [key: string]: unknown;
 }
 
-export interface ResolvedField extends Field {
-  tag: string;
-  label: string;
-  fields?: ResolvedField[];
-  details?: Record<string, any>;
-  applyChanges(
-    data: Data,
-    changes: Data,
-    field: ResolvedField,
-    document: Document,
-    content: CMSContent,
-  ): void | Promise<void>;
-}
+/**
+ * Contains the mapping of field types to the type of their value property.
+ */
+type FieldTypeToValueTypeMap = Prettify<
+  {
+    [K in FieldKeys]:
+      Extract<FieldTypeToPropertySelectionMap[K], "value"> extends never ? never
+        : unknown;
+  } & {
+    "checkbox": boolean;
+    "number": number;
+  }
+>;
+
+/**
+ * Creates the field options for one of the built-in field types with type `K`.
+ */
+type BuiltInField<K extends FieldKeys> =
+  & {
+    type: K;
+  }
+  & (FieldTypeToValueTypeMap[K] extends never ? {}
+    : {
+      value?: FieldTypeToValueTypeMap[K];
+    })
+  & Pick<FieldProperties, Exclude<FieldTypeToPropertySelectionMap[K], "value">>;
+
+/**
+ * Represents the options for a custom field type.
+ */
+type CustomField = Prettify<
+  & {
+    type: string & Record<never, never>;
+    //                ^ Typescript hack to suggest the correct keys but allow any string
+    //                  https://x.com/diegohaz/status/1524257274012876801
+    [key: string]: unknown;
+  }
+  & FieldProperties
+>;
+
+/**
+ * Represents the options for a field (both built in and custom).
+ */
+export type Field = Prettify<
+  | {
+    [K in FieldKeys]: BuiltInField<K>;
+  }[FieldKeys]
+  | CustomField
+>;
+
+/**
+ * Matches a string of form `/^.*:\s?.*!?$/` where the first part is the field name and the second part is the field type.
+ */
+export type FieldString = `${string}:${"" | " "}${Field["type"]}${"" | "!"}`;
```

Comment:
I didn't know Typescript could check string formats, this is great!

My only concern is if the tooltip text is helpful or not. This is what I see:

```
Type '"site textarea"' is not assignable to type 'Field | `${string}:number` | `${string}:number!` | `${string}:object` | `${string}:object!` | `${string}:text` | `${string}:text!` | `${string}:textarea` | `${string}:textarea!` | `${string}:markdown` | `${string}:markdown!` | `${string}:code` | `${string}:code!` | `${string}:datetime` | `${string}:datetime!` | `${s...'.deno-ts(2322)
```

Due it accepts too many options (and even field types not included by default), I'm not sure if this is good idea. Maybe this could be enough?

```ts
export type FieldString = `${string}:${string}${"" | "!"}`;
```

Reply:
Yeah that's reasonable. I had hoped to generate the `FieldKey` type based on what strings are passed into the `cms.field` function so that there wouldn't be a need to accommodate custom fields with the very general `string` type. I might take a crack at that next if I have time. I'm not sure that it's particularly useful if type can be any string. Also the autocompletion probably would be helped with a redesign of the api. Like I could image using tagged template strings to generate the field objects in place like `type\`name ("" | "!")\`` (ex. `text\`title!\``) or something, or using a tuple of the form `[name, type ("" | "!")]` (ex. `["title","text!"]`). Not saying that either of those changes are even slightly needed but just putting my thoughts down on paper.

Reply:
I see. The reason of using strings `"title: text!"` is for brevity and ergonomic. Tagged template strings is an interesting idea, but it would require to import the template function so we lost ergonomic. The use of arrays is also interesting but not so ergonomic as a simple string.

Maybe, people wanting for strict types should just use the object signature.